ghc-lib-gen/src/Ghclibgen.hs

```diff
@@ -283,127 +284,101 @@ ghcLibExtraFiles ghcFlavor =
-  (== ExitSuccess) . fst <$> systemOutput "bash -c \"ghc-pkg list | grep semaphore-compat\""
-
 setupModuleDepsPlaceholders :: GhcFlavor -> IO ()
 setupModuleDepsPlaceholders _ = do
-  forM_ [
-      "compiler/"
-    , "libraries/ghc-heap/"
-    , "libraries/ghc-internal/"
-    , "libraries/ghci/"
-    ] $ \path ->
-          whenM (doesDirectoryExist path) $
-             genPlaceholderModules path
-  forM_ [
-      "compiler/"
-    , "libraries/ghc-internal/src/"
-    ] $ \path ->
-          whenM (doesDirectoryExist path) $ do
-            files <- filter ((`elem` [".hs-boot"]) . takeExtension) <$> listFilesRecursive path
-            forM_ files $ \file -> do
-              let p = fromJust (stripPrefix path file)
-                  new_p = placeholderModulesDir </> p
-                  dir = System.FilePath.takeDirectory new_p
-              createDirectoryIfMissing True dir
-              copyFile file new_p
-
-calcModuleDeps :: [FilePath] -> [FilePath] -> [FilePath] -> GhcFlavor -> String -> Bool -> String -> String
-calcModuleDeps includeDirs _hsSrcDirs hsSrcIncludes ghcFlavor ghcPackagePath semaphoreCompatBootExists ghcMakeModeOutputFile = do
-  let series = ghcSeries ghcFlavor
-      rootModulePath = placeholderModulesDir </> "Main.hs"
-      cmd = unwords $
-        [ "ghc" ] ++
-        [ "-optP -DGHCI" | series < GHC_8_10 ] ++
-        [ "-optP -DSTAGE=2" | series < GHC_8_10 ] ++
-        [ "-optP -DGHC_IN_GHCI" | series < GHC_9_2 ] ++
-        [ "-dep-suffix ''"
-        , "-dep-makefile " ++ ghcMakeModeOutputFile
-        , "-M"
-        ] ++
-        includeDirs ++
-        [ "-ignore-package ghc"
-        , "-ignore-package ghci"
-        , "-package base"
-        ] ++
-        [ "-package exceptions" | series == GHC_9_0 ] ++
-        [ "-package semaphore-compat" | series >= GHC_9_8 ] ++
-        hsSrcIncludes ++
-        [ rootModulePath ]
-      cmd' =
-        if series < GHC_9_8 || semaphoreCompatBootExists then
-          cmd
-        else
-          "bash -c \"" ++ ghcPackagePath ++ " " ++ cmd ++ "\""
-    in cmd'
+  forM_
+    [ "compiler/",
+      "libraries/ghc-heap/",
+      "libraries/ghc-internal/",
+      "libraries/ghci/"
+    ]
+    $ \path ->
+      whenM (doesDirectoryExist path) $
+        genPlaceholderModules path
+  forM_
+    [ "compiler/",
+      "libraries/ghc-internal/src/"
+    ]
+    $ \path ->
+      whenM (doesDirectoryExist path) $ do
+        files <- filter ((`elem` [".hs-boot"]) . takeExtension) <$> listFilesRecursive path
+        forM_ files $ \file -> do
+          let p = fromJust (stripPrefix path file)
+              new_p = placeholderModulesDir </> p
+              dir = System.FilePath.takeDirectory new_p
+          createDirectoryIfMissing True dir
+          copyFile file new_p
+
+calcModuleDeps :: [FilePath] -> [FilePath] -> [FilePath] -> GhcFlavor -> FilePath -> String -> String
+calcModuleDeps includeDirs _hsSrcDirs hsSrcIncludes ghcFlavor cabalPackageDb ghcMakeModeOutputFile =
```

Comment:
I see quite a few options got removed here, why are they not needed anymore?

Reply:
this is the exciting part for me. the fact is that over the course of these last 5 years through all the versions we've supported and the complexity of the procedure as conducted via stack we accumulated a lot of workarounds and hacks that we determined were necessary for the time. now with stack removed and reliance strictly on the cabal solver i've gone back and sought opportunities to remove these things. this is the result!

Reply:
to be more explicit about what we've removed exactly here though: the preprocessor constants `STAGE=2`, `GHC_IN_GHCI` don't influence the outcome of the `ghc -M`. they probably never did and have always been redundant. the constant `GHCI` determines if `ghc -M` includes `GHCi.Run` and since the absence of that module does not affect the behavior of `ghc-lib-mini-compile-test` i judge it ok to exclude that module and drop the constant (less code is better).  the `ignore-package` directives were found to be unnecessary across all combinations tested so far (!) and so could be removed, similarly the `-package exceptions` has yet to be proven to be needed there (perhaps there's a config we haven't encountered yet that will require adding it back - we'll see). the `semaphoreCompatBootExists`  and `ghcPackagePath` related code got refactored into a better form using the `clear-package-db`, ..., `package-db` flags.

Reply:
Oh nice! That is exciting! Happy to see simplifications here :)